### PR TITLE
add elastic scaling MVP guide

### PR DESCRIPTION
**Pull request by alindima**
Resolves https://github.com/paritytech/polkadot-sdk/issues/4468

Gives instructions on how to enable elastic scaling MVP to parachain teams.

Still a draft because it depends on further changes we make to the slot-based collator: https://github.com/paritytech/polkadot-sdk/pull/4097

Parachains cannot use this yet because the collator was not released and no relay chain network has been configured for elastic scaling yet

### CODE REVIEW

File: docs/sdk/src/guides/enable_elastic_scaling_mvp.rs
Comment (sandreim) [inline docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:24]: IMO the limitation is due to the fact we are targeting 2s block times, more cores means lower than this number.

File: docs/sdk/src/guides/enable_elastic_scaling_mvp.rs
Comment (alindima) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:75]: This section is targeted at paras that use the cumulus template, not polkadot-parachain. If they want to switch to using elastic scaling they can't just use a CLI parameter, they still need to do these steps.

Once the slot based collator is merged I'll add a section here for parachains utilising `polkadot-parachain` binary

File: docs/sdk/src/guides/enable_elastic_scaling_mvp.rs
Comment (alindima) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:104]: yes they can. I specifically say how they all relate to each other and the formulas to derive them. I think the parachain teams can decide how to code them, this is just an example

File: docs/sdk/src/guides/enable_elastic_scaling_mvp.rs
Comment (alindima) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:102]: the outcome is the same as deciding on the maximum velocity

File: docs/sdk/src/guides/enable_elastic_scaling_mvp.rs
Comment (alindima) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:27]: yeah I created this guide assuming parachains that want to use multiple cores would do so to achieve higher throughput, but it can be also used to achieve lower latency (at least to inclusion in a candidate). I'll rephrase

File: docs/sdk/src/guides/enable_elastic_scaling_mvp.rs
Comment (alindima) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:131]: I didn't see any open issue for it. Is there one? Have we ever confirmed what the bug is?

The relay chain also adds +1 to the `max_candidate_depth`, so in reality this is no different from that value

File: docs/sdk/src/guides/enable_elastic_scaling_mvp.rs
Comment (alindima) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:110]: I added the constant computations based on maximum velocity

File: docs/sdk/src/guides/enable_elastic_scaling_mvp.rs
Comment (kianenigma) [inline docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:5]: You can also link to https://github.com/paritytech/polkadot-sdk/pull/4363 once it is merged. 

Moreover, I think you can also benefit a bit from the suggestiosn similar to https://github.com/paritytech/polkadot-sdk/pull/4363#issuecomment-2141171928

File: docs/sdk/src/guides/enable_elastic_scaling_mvp.rs
Comment (kianenigma) [inline docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:24]: How do we know that these 3 para-blocks are still valid when imported in 3 parallel cores?

For example, there are 2 tx in each parablock. The collator proposes `[t1, t2, t3, t4, t5, t6]` and they are all valid. But the validity of `t6` depends on the execution of `t1`. When imported in 3 cores, `t1` and `t6` are no longer present. 

In general, I would assume all of this to be fixed in the cumulus block building code. My question is, does it?

File: docs/sdk/src/guides/enable_elastic_scaling_mvp.rs
Comment (kianenigma) [inline docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:71]: This should use `docify` into `parachain-template`. Possibly, it will be hard to do it, and I belive this is a good thing, as it will force us to simplify the API a bit. Give it a try please. As noted above, it is similar to the suggestions in https://github.com/paritytech/polkadot-sdk/pull/4363#issuecomment-2141171928

Comment (kianenigma) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:71]: For example, my worry with this doc rn is that the moment we change anything in the parachain template, it will likely become invalid.

File: docs/sdk/src/guides/enable_elastic_scaling_mvp.rs
Comment (kianenigma) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:81]: Guides should be built for `template/parachain`, not `polkadot-parachain` IMO. 

That being said, as noted above, I think this doc can be written better if it reuses the code of `template/parachain`

File: docs/sdk/src/guides/enable_elastic_scaling_mvp.rs
Comment (kianenigma) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:110]: I definitely have seen this velocity stuff in the async backing guides PR as well, I think it is best to first push that to completion in the best possible shape, then build this on top of it.

File: docs/sdk/src/guides/enable_elastic_scaling_mvp.rs
Comment (kianenigma) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:125]: use docify please :)

**Review by kianenigma — CHANGES_REQUESTED** [review]
It seems best to me for you to first coordinate with https://github.com/paritytech/polkadot-sdk/pull/4363, possily push it to completion of Radha is not available, then build this on top of.

File: docs/sdk/src/guides/enable_elastic_scaling_mvp.rs
Comment (sandreim) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:24]: These 3 blocks are expected to form a chain, the ones that don't will not be included.

File: docs/sdk/src/guides/enable_elastic_scaling_mvp.rs
Comment (alindima) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:71]: What you're saying makes sense, but at least part of this guide is for parachains that are already running and want to update. The default template will already be elastic-scaling ready, much like it is already async backing ready.
Still, the document for updating to async backing was targeted at existing parachains (the ones that start from the template barely need to do anything).

Comment (alindima) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:71]: maybe it would be best to first update the template and then just reference it in the doc as an example, instead of giving find and replace instructions. however, the feature is not yet stable enough so that we can update the template. therefore I'll leave the PR as it is for now (that's also why it's a draft)

Comment (alindima) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:71]: so basically there are two conlicting things:
1. the document should help existing parachains upgrade
2. keeping the document up to date with the template

If we want 2, the document will lack all the changes needed for 1

File: docs/sdk/src/guides/enable_elastic_scaling_mvp.rs
Comment (alindima) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:24]: > These 3 blocks are expected to form a chain, the ones that don't will not be included.

yes, also a candidate will not be included until all of its ancestors are included. if one ancestor is not included (times out availability) or is concluded invalid via a dispute, all of its descendants will also be evicted from the cores. So we only deal with candidate chains

**Comment by DrW3RK** [comment]
Applied changes to https://github.com/paritytech/polkadot-sdk/pull/4363 Needs one more approval for merge

**Comment by kianenigma** [comment]
> Applied changes to #4363 Needs one more approval for merge

Almost, but not [yet](https://github.com/paritytech/polkadot-sdk/pull/4363#issuecomment-2164260859) :)

File: docs/sdk/src/guides/enable_elastic_scaling_mvp.rs
Comment (eskimor) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs]: > higher throughput,

can also mean more data.

**Comment by kianenigma** [comment]
Would be happy to review this once it is updated based on the previous guides.

**Comment by alindima** [comment]
> Would be happy to review this once it is updated based on the previous guides.

Thanks, I'll let you know when this is ready. I've put this on hold a bit waiting for https://github.com/paritytech/polkadot-sdk/pull/4097 to be merged (and it has been) and getting more specific performance numbers here: https://github.com/paritytech/polkadot-sdk/issues/4696

**Comment by alindima** [comment]
Ok I thought about using docify here. But how can I, considering that the parachain template is not updated for elastic scaling yet? (and we don't plan to update it yet as it's still an experimental MVP).
AFAICT I can only reference existing code which is not really helpful. We can use docify only after we modify the template, by which time the async backing guide will no longer be able to use docify. @kianenigma

**Comment by alindima** [comment]
another problem I see with docify is that it only works on types. you cannot annotate and import blocks of code

**Comment by alindima** [comment]
I tried using docify as much as it made sense. Please review

**Review by kianenigma — APPROVED** [review]

**Review by sandreim — APPROVED** [review]

File: docs/sdk/src/guides/enable_elastic_scaling_mvp.rs
Comment (kianenigma) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:29]: Sorry, I still don't get this. 

@sandreim if they form a chain, and part of the chain is executed in one core and part of it in another core, how does either of the cores check that the whole thing is a chain? 

in my example, `[t1, t2, t3, t4, t5, t6]`, `[t1, t2, t3]` goes into one core, `[t4, t5, t6]` into another. The whole `[t1 -> t6]` indeed forms a chain, and execution of t5 depends on the execution of t2. 

Perhaps what you mean to say is that the transactions that go into different cores must in fact be independent of one another?

Comment (sandreim) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:29]: The transactions are not independent. We achieve parallel execution even in that case, and still check they form a chain by passing in the appropriate validation inputs (https://github.com/paritytech/polkadot-sdk/blob/d1979d4be9c44d06c62aeea44d30f5f201d67b5e/polkadot/primitives/src/v7/mod.rs#L661) . We can validate t2 because we already have the parent head data of t1 from the collator of t2. So we can correctly construct the inputs and the PoV contains the right data ( t2 was built after t1 by the collator).

Does it makes sense.l?

Comment (alindima) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:29]: > [t1, t2, t3] goes into one core, [t4, t5, t6] into another.

This is not really possible. If a parachain uses multiple cores, all the blocks occupying the cores will form a chain.
In your example, on core 1, we'll probably have [t1,t3,t5] and on core 2 [t2,t4, t6].

Comment (kianenigma) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:29]: > > [t1, t2, t3] goes into one core, [t4, t5, t6] into another.
> 
> This is not really possible. If a parachain uses multiple cores, all the blocks occupying the cores will form a chain. In your example, on core 1, we'll probably have [t1,t3,t5] and on core 2 [t2,t4, t6].

I don't think you can generalize, because you are assuming everyone will use a specific collator implementation, right? I have leases two cores, and I run their collators, and therefore it is me who chose to which cores I send which transactions. 

Let's speak only from the perspective of Polkadot, and its cores here. For all I know, collators are written by arbitrary external teams, in much the same way that ETH block builders are separate from core validators. 

From the perspective of Polkadot, when the same parachain is using two cores, what is the assumptions?

> We can validate t2 because we already have the parent head data of t1 from the collator of t2. So we can correctly construct the inputs and the PoV contains the right data ( t2 was built after t1 by the collator).

You are also talking about the collators here, while I am curious to know how does two different Polkadot cores (and therefore two different backing groups) simultaneously validate two blocks that might have inter-dependence.

Comment (kianenigma) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:29]: Is this the answer? 

`[t1, t2, t3]` goes into one core, `[t4, t5, t6]`, but the PoV of the latter contains the full execution of the former? 

I think this is fine, but truthfully to scale up, I think different transactions going into different cores must be independent, or else the system can only scale as much as you can jack up one collator.

Comment (ordian) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:29]: > but the PoV of the latter contains the full execution of the former?

PoV of `[t4, t5, t6]` would refer to the state post `[t1, t2, t3]` execution.

> I think different transactions going into different cores must be independent, or else the system can only scale as much as you can jack up one collator

One way of achieving that would be to have a DAG instead of a blockchain (two blocks having the same parent state). But then you'd need to somehow ensure they are truly independent. This could be done with e.g. specifying dependencies in the transactions themselves (a ala Solana or Ethereum access lists).

Another way would be to implement a collator differently with optimistic concurrency control (ala Monad). This only requires modification on the collator side and does not affect transaction format.

Comment (kianenigma) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:29]: Okay, thanks @ordian. 

I totally agree with all of your directions as well. I am not sure if you have seen it or not, but my MSC Thesis was on the same topic 🙈 https://github.com/kianenigma/SonicChain. I think what I have done here is similar to access list, and it should be quite easy to add to FRAME and Substrate: each tx to declare, via its code author, what storage keys it "thinks" it will access. Then the collators can easily agree among themselves to collate non-conflicting transactions. 

This is a problem that is best solve from the collator side, and once there is a lot of demand. Polkadot is already doing what it should do, and should not do any "magic" to handle this. 

Once there is more demand: 

1. Either collators just jack up, as they kinda are expected to do now. This won't scale a lot but it will for a bit. 
2. I think the access list stuff is super cool and will scale
3. OCC is fancy but similarly doesn't scale, because there is only so many CPU cores, and you are still bound to one collator somehow filling up 8 Polkadot cores. Option 2 is much more powerful, because you can enable 8 collators to fill 8 blocks simultaneously.

Comment (sandreim) [reply docs/sdk/src/guides/enable_elastic_scaling_mvp.rs:29]: Current status quo is that we rely on 1 (beefy collators). 2 for sure is something that can scale well, but it seems to be complicated and is not really compatible with the relay chain which expects chains not a DAG. https://github.com/paritytech/polkadot-sdk/issues/4696#issuecomment-2173250798 shows how the limitations of what is possible with ref hw and 5 collators.

We did a nice brainstorming session with @skunert and @eskimor on the subject some time ago. We think that best way to go forward is to implement a transaction streaming mechanism. At the begging of each slot, the block author sends transactions to the next block author as it pushes them in the current block. By the time it announces the block, the next author should already have all state changes applied and doesn't need to wait to import it and can immediately start building his own block. And so on. 

If that is not enough, next block author can start to speculatively build the it's next block and add/remove transactions as it learns what the current author is putting in his blocks.

**Pull request merged**
